### ISSUE TITLE
Autogeneration not generating only changes to model

### ISSUE THREAD

**Issue by okyere** [miguelgrinberg/Flask-Migrate#93]
I followed the documentation and I'm able to do my first migration with the current state of models. Doing 'db upgrade' does creates the corresponding tables on my schema for me. That's all good. From the documentation, I was under the impression that if I make changes to my models eg. add a column, add a new table, etc and then I do 'db migrate', flask-migrate will generate upgrade and downgrade scripts for only the things that have changed since the last migration. Basically, it would compare the current state of the schema to the state of the models and only generate scripts for what has CHANGED. Is that correct?
Because what I see is that, after I change the model and run 'db migrate', the upgrade script contains code that creates ALL the tables and the downgrade code to drop all the tables. I was expecting that it will only be the newly added or modified stuff that will be reflected in the autogenerated script. For this reason, when I go ahead to do 'db upgrade' , I get:

sqlalchemy.exc.DatabaseError: (cx_Oracle.DatabaseError) ORA-00955: name is already used by an existing object

Basically, the upgrade method is trying to recreate all the tables in the schema instead of only creating new things that have been added to the models or adding things that have been added to the models.
Am I misunderstanding something? Am I supposed to manually go and remove the code that recreates existing objects in the schema before running the ''db upgrade"? Thanks.

**Comment by miguelgrinberg** [comment]
@okyere the `db upgrade` command should only pick up the differences between the current state of your database and the current state of your models. So clearly there is something that isn't working as expected in your environment. Would you care to add the complete console output of the first and second migrations, along with a copy of your model(s) definition for each of them?

**Comment by okyere** [comment]
Console output for first and second migrations:
(uploadenv)Nanas-MacBook-Pro:data_upload_app okyere$ python manage.py db init
  Creating directory /Users/okyere/Repos/data_upload_app/migrations ... done
  Creating directory /Users/okyere/Repos/data_upload_app/migrations/versions ... done
  Generating /Users/okyere/Repos/data_upload_app/migrations/alembic.ini ... done
  Generating /Users/okyere/Repos/data_upload_app/migrations/env.py ... done
  Generating /Users/okyere/Repos/data_upload_app/migrations/README ... done
  Generating /Users/okyere/Repos/data_upload_app/migrations/script.py.mako ... done
  Please edit configuration/connection/logging settings in '/Users/okyere/Repos/data_upload_app/migrations/alembic.ini'
  before proceeding.
(uploadenv)Nanas-MacBook-Pro:data_upload_app okyere$ python manage.py db migrate
INFO  [alembic.runtime.migration] Context impl OracleImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.autogenerate.compare] Detected added table 'dut_departments'
INFO  [alembic.autogenerate.compare] Detected added table 'dut_field_types'
INFO  [alembic.autogenerate.compare] Detected added table 'dut_uploadslog'
INFO  [alembic.autogenerate.compare] Detected added table 'dut_tables_info'
INFO  [alembic.autogenerate.compare] Detected added table 'dut_fields'
  Generating /Users/okyere/Repos/data_upload_app/migrations/versions/1d83e77b9c6_.py ... done
(uploadenv)Nanas-MacBook-Pro:data_upload_app okyere$ python manage.py db upgrade
INFO  [alembic.runtime.migration] Context impl OracleImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> 1d83e77b9c6, empty message
(uploadenv)Nanas-MacBook-Pro:data_upload_app okyere$ python manage.py db migrate
INFO  [alembic.runtime.migration] Context impl OracleImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.autogenerate.compare] Detected added table 'dut_departments'
INFO  [alembic.autogenerate.compare] Detected added table 'dut_field_types'
INFO  [alembic.autogenerate.compare] Detected added table 'dut_uploadslog'
INFO  [alembic.autogenerate.compare] Detected added table 'dut_tables_info'
INFO  [alembic.autogenerate.compare] Detected added table 'dut_fields'
  Generating /Users/okyere/Repos/data_upload_app/migrations/versions/4959eff9961_.py ... done
(uploadenv)Nanas-MacBook-Pro:data_upload_app okyere$ python manage.py db upgrade
INFO  [alembic.runtime.migration] Context impl OracleImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 1d83e77b9c6 -> 4959eff9961, empty message
Traceback (most recent call last):
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/engine/base.py", line 1139, in _execute_context
    context)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/engine/default.py", line 450, in do_execute
    cursor.execute(statement, parameters)
cx_Oracle.DatabaseError: ORA-00955: name is already used by an existing object


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "manage.py", line 16, in <module>
    manager.run()
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/flask_script/__init__.py", line 412, in run
    result = self.handle(sys.argv[0], sys.argv[1:])
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/flask_script/__init__.py", line 383, in handle
    res = handle(*args, **config)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/flask_script/commands.py", line 216, in __call__
    return self.run(*args, **kwargs)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/flask_migrate/__init__.py", line 216, in upgrade
    command.upgrade(config, revision, sql=sql, tag=tag)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/command.py", line 174, in upgrade
    script.run_env()
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/script/base.py", line 397, in run_env
    util.load_python_file(self.dir, 'env.py')
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/util/pyfiles.py", line 81, in load_python_file
    module = load_module_py(module_id, path)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/util/compat.py", line 68, in load_module_py
    module_id, path).load_module(module_id)
  File "<frozen importlib._bootstrap>", line 539, in _check_name_wrapper
  File "<frozen importlib._bootstrap>", line 1614, in load_module
  File "<frozen importlib._bootstrap>", line 596, in _load_module_shim
  File "<frozen importlib._bootstrap>", line 1220, in load
  File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1129, in _exec
  File "<frozen importlib._bootstrap>", line 1471, in exec_module
  File "<frozen importlib._bootstrap>", line 321, in _call_with_frames_removed
  File "migrations/env.py", line 87, in <module>
    run_migrations_online()
  File "migrations/env.py", line 80, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/runtime/environment.py", line 797, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/runtime/migration.py", line 312, in run_migrations
    step.migration_fn(**kw)
  File "/Users/okyere/Repos/data_upload_app/migrations/versions/4959eff9961_.py", line 23, in upgrade
    sa.UniqueConstraint('name')
  File "<string>", line 8, in create_table
  File "<string>", line 3, in create_table
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/operations/ops.py", line 1098, in create_table
    return operations.invoke(op)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/operations/base.py", line 318, in invoke
    return fn(self, operation)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/operations/toimpl.py", line 101, in create_table
    operations.impl.create_table(table)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/ddl/impl.py", line 194, in create_table
    self._exec(schema.CreateTable(table))
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/ddl/oracle.py", line 22, in _exec
    result = super(OracleImpl, self)._exec(construct, *args, **kw)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/alembic/ddl/impl.py", line 118, in _exec
    return conn.execute(construct, *multiparams, **params)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/engine/base.py", line 914, in execute
    return meth(self, multiparams, params)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/sql/ddl.py", line 68, in _execute_on_connection
    return connection._execute_ddl(self, multiparams, params)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/engine/base.py", line 968, in _execute_ddl
    compiled
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/engine/base.py", line 1146, in _execute_context
    context)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/engine/base.py", line 1341, in _handle_dbapi_exception
    exc_info
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/util/compat.py", line 188, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=exc_value)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/util/compat.py", line 181, in reraise
    raise value.with_traceback(tb)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/engine/base.py", line 1139, in _execute_context
    context)
  File "/Users/okyere/miniconda3/envs/uploadenv/lib/python3.4/site-packages/sqlalchemy/engine/default.py", line 450, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.DatabaseError: (cx_Oracle.DatabaseError) ORA-00955: name is already used by an existing object
 [SQL: '\nCREATE TABLE dut_departments (\n\tid INTEGER NOT NULL, \n\tname VARCHAR2(120 CHAR) NOT NULL, \n\tPRIMARY KEY (id), \n\tUNIQUE (name)\n)\n\n']

models for first migration:
===================


from app import db, theapp
from sqlalchemy.schema import Sequence, CreateSequence
import datetime

def _create_sequence(name):
    # if db is Oracle create the sequence
    if db.engine.name.upper() == 'ORACLE':
        query = "SELECT * FROM user_sequences WHERE sequence_name = '{}'".format(name.upper())
        result = db.session.execute(query)

        if not result.scalar():
            db.session.execute('CREATE SEQUENCE {}'.format(name.upper()))
            db.session.commit()

def _drop_sequence(name):
    # if db is Oracle create the sequence
    if db.engine.name.upper() == 'ORACLE':
        sql = 'DROP SEQUENCE {}'.format(name.upper())
        db.session.execute(sql)
        db.session.commit()


class FieldType(db.Model): # one to field
    __tablename__ = 'dut_field_types'

    id = db.Column(db.Integer, db.Sequence('DUT_FIELD_TYPES_SEQ'), primary_key = True)
    name = db.Column(db.String(120), unique=True, nullable=False)
    python_type = db.Column(db.String(120))
    python_import = db.Column(db.String(120))
    fields = db.relationship("Field", backref = "type", lazy = 'dynamic')

    _create_sequence('DUT_FIELD_TYPES_SEQ')

    def __str__(self):
        return self.name

def possible_types():
    return FieldType.query.order_by("id")

class Field(db.Model): # many to field type
    __tablename__ = 'dut_fields'
    id = db.Column(db.Integer, db.Sequence('DUT_FIELDS_SEQ'), primary_key = True)
    name = db.Column(db.String(120), nullable = False)

    fieldtype_id = db.Column(db.Integer, db.ForeignKey("dut_field_types.id"))
    tableinfo_id = db.Column(db.Integer, db.ForeignKey('dut_tables_info.id'))

    _create_sequence('DUT_FIELDS_SEQ')

    def __str__(self):
        return self.name

class TableInfo(db.Model):
    __tablename__ = 'dut_tables_info'

    id = db.Column(db.Integer, db.Sequence('DUT_TABLES_INFO_SEQ'), primary_key = True)
    descriptive_name = db.Column(db.String(120), unique = True, nullable=False) # name for user to identify table
    table_name = db.Column(db.String(30), unique = True, nullable=False) # database table name
    fields = db.relationship('Field', backref='table_info', lazy='dynamic')
    department_id = db.Column(db.Integer, db.ForeignKey('dut_departments.id'))

    _create_sequence('DUT_TABLES_INFO_SEQ')

    def __str__(self):
        return self.descriptive_name

def possible_tableinfos():
    return TableInfo.query.order_by("descriptive_name asc")

class Department(db.Model):
    __tablename__ = 'dut_departments'

    id = db.Column(db.Integer, db.Sequence('DUT_DEPARTMENTS_SEQ'), primary_key=True)
    name = db.Column(db.String(120), unique=True, nullable=False)

    # one-to-many relationship (a Department can have several associated TableInfos)
    tables_info = db.relationship('TableInfo', backref = 'department', lazy='joined')

    _create_sequence('DUT_DEPARTMENTS_SEQ')

    def __str__(self):
        return self.name

class UploadsLog(db.Model):
    __tablename__ = 'dut_uploadslog'

    id = db.Column(db.Integer, db.Sequence('DUT_UPLOADSLOG_SEQ'), primary_key=True)
    user_name = db.Column(db.String(120), nullable=False)
    user_cwsid = db.Column(db.String(120), nullable=False)
    file_name = db.Column(db.String(120), nullable=False)
    destination_table = db.Column(db.String(120), nullable=False)
    timestamp = db.Column(db.TIMESTAMP, nullable=False)


    _create_sequence('DUT_UPLOADSLOG_SEQ')

def possible_departments():
    return Department.query



models.py for second migration:
==========================
from app import db, theapp
from sqlalchemy.schema import Sequence, CreateSequence
import datetime

def _create_sequence(name):
    # if db is Oracle create the sequence
    if db.engine.name.upper() == 'ORACLE':
        query = "SELECT * FROM user_sequences WHERE sequence_name = '{}'".format(name.upper())
        result = db.session.execute(query)

        if not result.scalar():
            db.session.execute('CREATE SEQUENCE {}'.format(name.upper()))
            db.session.commit()

def _drop_sequence(name):
    # if db is Oracle create the sequence
    if db.engine.name.upper() == 'ORACLE':
        sql = 'DROP SEQUENCE {}'.format(name.upper())
        db.session.execute(sql)
        db.session.commit()


class FieldType(db.Model): # one to field
    __tablename__ = 'dut_field_types'

    id = db.Column(db.Integer, db.Sequence('DUT_FIELD_TYPES_SEQ'), primary_key = True)
    name = db.Column(db.String(120), unique=True, nullable=False)
    python_type = db.Column(db.String(120))
    python_import = db.Column(db.String(120))
    fields = db.relationship("Field", backref = "type", lazy = 'dynamic')

    _create_sequence('DUT_FIELD_TYPES_SEQ')

    def __str__(self):
        return self.name

def possible_types():
    return FieldType.query.order_by("id")

class Field(db.Model): # many to field type
    __tablename__ = 'dut_fields'
    id = db.Column(db.Integer, db.Sequence('DUT_FIELDS_SEQ'), primary_key = True)
    name = db.Column(db.String(120), nullable = False)

    fieldtype_id = db.Column(db.Integer, db.ForeignKey("dut_field_types.id"))
    tableinfo_id = db.Column(db.Integer, db.ForeignKey('dut_tables_info.id'))

    _create_sequence('DUT_FIELDS_SEQ')

    def __str__(self):
        return self.name

class TableInfo(db.Model):
    __tablename__ = 'dut_tables_info'

    id = db.Column(db.Integer, db.Sequence('DUT_TABLES_INFO_SEQ'), primary_key = True)
    descriptive_name = db.Column(db.String(120), unique = True, nullable=False) # name for user to identify table
    table_name = db.Column(db.String(30), unique = True, nullable=False) # database table name
    fields = db.relationship('Field', backref='table_info', lazy='dynamic')
    department_id = db.Column(db.Integer, db.ForeignKey('dut_departments.id'))

    _create_sequence('DUT_TABLES_INFO_SEQ')

    def __str__(self):
        return self.descriptive_name

def possible_tableinfos():
    return TableInfo.query.order_by("descriptive_name asc")

class Department(db.Model):
    __tablename__ = 'dut_departments'

    id = db.Column(db.Integer, db.Sequence('DUT_DEPARTMENTS_SEQ'), primary_key=True)
    name = db.Column(db.String(120), unique=True, nullable=False)

    # one-to-many relationship (a Department can have several associated TableInfos)
    tables_info = db.relationship('TableInfo', backref = 'department', lazy='joined')

    _create_sequence('DUT_DEPARTMENTS_SEQ')

    def __str__(self):
        return self.name

class UploadsLog(db.Model):
    __tablename__ = 'dut_uploadslog'

    id = db.Column(db.Integer, db.Sequence('DUT_UPLOADSLOG_SEQ'), primary_key=True)
    user_name = db.Column(db.String(120), nullable=False)
    user_cwsid = db.Column(db.String(120), nullable=False)
    last_name = db.Column(db.String(120))
    file_name = db.Column(db.String(120), nullable=False)
    destination_table = db.Column(db.String(120), nullable=False)
    timestamp = db.Column(db.TIMESTAMP, nullable=False)


    _create_sequence('DUT_UPLOADSLOG_SEQ')

def possible_departments():
    return Department.query

To keep this simple the difference between the first and second migrations is the addition of ::
last_name = db.Column(db.String(120))

to class UploadsLog(db.Model) . I was hoping that  the migration will only pick up that difference. Please help if you can.

**Comment by okyere** [comment]
[console_output.txt](https://github.com/miguelgrinberg/Flask-Migrate/files/75369/console_output.txt)

[models_1.txt](https://github.com/miguelgrinberg/Flask-Migrate/files/75370/models_1.txt)
[models_2.txt](https://github.com/miguelgrinberg/Flask-Migrate/files/75371/models_2.txt)

The only difference between models 1 and 2 is the addition of last_name column on line 92 in models 2. The console output has the initial init, first migrate and upgrade; and second migrate and upgrade.

**Comment by miguelgrinberg** [comment]
Thanks. I don't see anything odd in these files, it seems that for the purposes of generating automatic migrations Alembic is always starting from an empty db, but when running a migration everything works fine.

Have you modified any of the configuration files in the `migrations` directory? A bad `env.py`, for example, could cause this behavior, I suppose.

**Comment by okyere** [comment]
My observation exactly. I don't know why alembic is always starting from an empty db even when objects already exist. No, I don't modify files under the migration folder. I have only looked at the env.py with limited understanding. But I don't edit anything. If it helps, I'm attaching the manage.py and the __init__.py under the 'app' folder. If anything looks odd, let me know. The app is structured based on your mega tutorial. That said, thanks for your contribution to the this community.

[__init__.txt](https://github.com/miguelgrinberg/Flask-Migrate/files/75741/__init__.txt)
[manage.txt](https://github.com/miguelgrinberg/Flask-Migrate/files/75742/manage.txt)

**Comment by miguelgrinberg** [comment]
Did you try another database with your app? Switching to sqlite should not take any effort, just change the connection url. It would be interesting to see if the problem goes away on a different database.

**Comment by okyere** [comment]
I tried with sqlite and it worked. So I have more questions than answers. I've read through some docs for alembic and I cannot find anywhere it states that autogeneration is not supported for Oracle. I don't know how to move on from here.

**Comment by miguelgrinberg** [comment]
Hmm. Well, okay, at least we know it's not something in your environment that is causing this.

Know any way to get a short-term access to a Oracle database, without having to sell my soul to Oracle? I would like to try this, because I agree, Alembic docs or forums do not suggest this does not work, in fact, the contrary, people seem to use autogenerated migrations.

**Comment by okyere** [comment]
Thanks for your willingness to look into this. Don't sell your soul to Oracle! :-) Oracle XE is a free version for development, etc. If you've jumped on the docker bandwagon, here's a docker image with Oracle xe for your use: https://github.com/MaksymBilenko/docker-oracle-xe-11g . It is the cleanest way to use oracle without installing it directly on your OS. 
First install the oracle instant client for your OS http://www.oracle.com/technetwork/topics/winx64soft-089540.html, 
Then install cx_oracle in your virtualenv. 
The github repo link I gave above has the installation information for the oracle image and the credentials to log into the database.

Unfortunately, we use oracle at work so my learning/experimentations have all been with oracle. Hence this issue. If you have any issues getting oracle set up, please let me know. I spent hours on some of those things in my early days a few months ago. Thanks again.

**Comment by miguelgrinberg** [comment]
Okay, I'll give this a try and let you know. Thanks!

**Comment by okyere** [comment]
Miguel, I just found out why this is happening. Unless you still want to play with Oracle (No one does!), ignore this. I reached out to Mike on the alembic google groups about this : https://groups.google.com/d/msg/sqlalchemy-alembic/Q32ErOhqyiM/OA6bIzebDQAJ

As it turns out, it was a combination of Oracle craziness and my lack of knowledge of that craziness. I was using the 'SYSTEM' user and schema for development. That's a no no. I created a standard user account in oracle and tested again. Not it works. Apparently, user tables are not supposed to be created on system schema. Mike states that sqlalchemy SQLAlchemy's reflection queries explicitly skip over SYSTEM so that these administrative tables aren't included in table lists. So there it is. Thanks for your efforts and your continued support for the Flask community.

**Comment by miguelgrinberg** [comment]
Awesome, glad that went and asked the expert, it would have taken me a long time to figure this out myself.

**Comment by syedwaseemjan** [comment]
Hey Miguel, I am also facing something like this. The only difference is that I am using Postgresql. I am not generating my tables in default public schema instead I am creating two schemas to better arrange my tables. Do you have any idea why would that happen?
I have posted my question on stackoverflow too http://stackoverflow.com/questions/41150956/flask-migrate-not-generating-correct-changes
